src/expr/src/scalar/func/impls/mztimestamp.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use chrono::{DateTime, NaiveDateTime, Utc};
+
+use mz_ore::result::ResultExt;
+use mz_repr::{adt::numeric::Numeric, strconv, Timestamp};
+
+use crate::EvalError;
+
+// Conversions to MzTimestamp, and a single conversion from MzTimestamp to
+// String. In general we want to make MzTimestamp a more opaque type, so we
+// easily support casting things to it but not from it.
+
+sqlfunc!(
+    #[sqlname = "mz_timestamp_to_text"]
+    #[preserves_uniqueness = true]
+    fn cast_mz_timestamp_to_string(a: Timestamp) -> String {
+        let mut buf = String::new();
+        strconv::format_mztimestamp(&mut buf, a);
+        buf
+    }
+);
+
+sqlfunc!(
+    #[sqlname = "text_to_mz_timestamp"]
+    #[preserves_uniqueness = true]
+    fn cast_string_to_mz_timestamp(a: String) -> Result<Timestamp, EvalError> {
+        strconv::parse_mztimestamp(&a).err_into()
+    }
+);
+
+sqlfunc!(
+    #[sqlname = "numeric_to_mz_timestamp"]
+    #[preserves_uniqueness = true]
```

Comment:
Same as above; does this error if there are e.g. decimal values that would be truncated?

Reply:
This errors if the decimal value is truncated or otherwise doesn't exactly fit. Added a comment.